src/libcore/iter/traits.rs

```diff
@@ -581,13 +582,15 @@ pub trait Product<A = Self>: Sized {
     fn product<I: Iterator<Item=A>>(iter: I) -> Self;
 }
 
+// NB: explicitly use Add and Mul here to inherit overflow checks
 macro_rules! integer_sum_product {
     ($($a:ident)*) => ($(
         #[stable(feature = "iter_arith_traits", since = "1.12.0")]
         impl Sum for $a {
+            #[rustc_inherit_overflow_checks]
             fn sum<I: Iterator<Item=$a>>(iter: I) -> $a {
                 iter.fold(0, |a, b| {
-                    a.checked_add(b).expect("overflow in sum")
+                    Add::add(a, b)
```

Comment:
That hasn't been the case since some time before 1.0, at the type level, and since March or so this year as far as the low-level representation and optimizations go.